### PR TITLE
[DRAFT] [TTS] FastPitch training tutorial

### PR DESCRIPTION
**Pull request by subhankar-ghosh**
# Addition of FastPitch training tutorial

Please have one of the following:

* New feature: (short feature description) FastPitch training tutorial.
This PR adds FastPitch tutorial to the same notebook as MixerTTS as both generate spectrogram. This tutorial goes through the data download, preprocess, FastPitch import and training pipeline.
Tutorial was added with careful consideration to make minimal changes to MixerTTS tutorial.

## Before your PR is "Ready for review"
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or updated any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

**Review by Oktai15 — CHANGES_REQUESTED** [review]
@subhankar-ghosh, please, use new config for FastPitch + Aligner (https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/fastpitch_align_v1.05.yaml). Config which you use will be deprecated, because it uses old dataset.

**Comment by itzsimpl** [comment]
Where does one get these values from?
```python3
  model.pitch_avg=121.9 model.pitch_std=23.1 \
  model.pitch_fmin=30 model.pitch_fmax=512
```

Is 
```python3
  prior_folder=./Priors_an4 \
```
an empty folder? If not, how does one precompute the priors?

Since this is a tutorial, could you please provide further and more detailed explanations.

**Comment by blisc** [comment]
> Where does one get these values from?
> 
> ```python
>   model.pitch_avg=121.9 model.pitch_std=23.1 \
>   model.pitch_fmin=30 model.pitch_fmax=512
> ```

Are the explanations from https://github.com/NVIDIA/NeMo/blob/main/tutorials/tts/FastPitch_Finetuning.ipynb sufficient?

> Is
> 
> ```python
>   prior_folder=./Priors_an4 \
> ```
> 
> an empty folder? If not, how does one precompute the priors?
> 
> Since this is a tutorial, could you please provide further and more detailed explanations.

It is computed on the fly if it is empty or one can precompute them from https://github.com/NVIDIA/NeMo/blob/main/scripts/dataset_processing/tts/extract_sup_data.py.

@subhankar-ghosh Can you update the notebook to address these comments?

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 1]
```diff
@@ -148,10 +156,87 @@
...
+   "id": "d12262e7",
+   "metadata": {},
+   "source": [
+    "### FastPitch\n",
+    "\n",
+    "FastPitch is  a fully-parallel text-to-speech model based on FastSpeech, conditioned on fundamental frequency contours. The model predicts pitch contours during inference. Conditioning on frequency contours improves the overall quality of synthesized speech. For more details about model, please refer to the original [paper](https://arxiv.org/abs/2006.06873).\n",
+    "\n",
+    "FastPitch like most NeMo models is defined as a LightningModule.\n",
+    "\n",
+    "Let's take a look using NeMo's pretrained model and how to use it to generate spectrograms."
```
Comment (redoctopus) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:168]: May also be worth pointing out where the procedure differs from Mixer-TTS so users can better learn how to generalize to different models in the future

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 2]
```diff
@@ -315,17 +400,66 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "01e7ca12",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# download data and manifests\n",
+    "!wget https://github.com/NVIDIA/NeMo/releases/download/v0.11.0/test_data.tar.gz && mkdir -p tests/data && tar xzf test_data.tar.gz -C tests/data"
```
Comment (redoctopus) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:445]: Nitpick--could you split this line for better readability?

**Comment by itzsimpl** [comment]
I would prefer more information, or even better a script that is part of NeMo, whose sole function is to precompute the required parameters. Fixed and maintained with the rest of the repository. At the moment everything is very unclear. One needs to jump through hoops hunting for correct information. 

For example it is unclear how one can precompute the required pitch, priors and duration (in case of non fine-tuned FastPitch). As mentioned, there is no script to show how this is done. There is an open discussion on this topic (#3138), but it remains unanswered. The only reference to how the data is computed can be found in https://github.com/NVIDIA/NeMo/blob/9a1cc36de01bc557df7f540770bec919754e21ee/nemo/collections/asr/data/audio_to_text.py#L752

This references a PR from February 2021 https://github.com/NVIDIA/NeMo/pull/1799, which links to a script from DLE. However, since the PR, DLE has changed and the data preparation script (to which the PR points to) as well. Indeed the PR points directly to [prepare_dataset.sh](https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/SpeechSynthesis/FastPitch/scripts/prepare_dataset.sh), which is now based on [prepare_dataset.py](https://github.com/NVIDIA/DeepLearningExamples/blob/18e77561231c0a9f9f6a718cba2393260f71dc26/PyTorch/SpeechSynthesis/FastPitch/prepare_dataset.py), but before August 2021 and the introduction of FastPitch 1.1 (auto aligned version) it used to be based on [extract_mels.py]( https://github.com/NVIDIA/DeepLearningExamples/blob/afe00719cad16281130dc1c260f4575d0f9a8656/PyTorch/SpeechSynthesis/FastPitch/extract_mels.py). And this later script too has changed after February 2021.

If one compares the two scripts, there are ample differences: 
- `prepare_dataset.py` can be used to extract `mels`, `pitch` and `alignment_priors`, but not `durations` (required by [FastPitch.py](https://github.com/NVIDIA/NeMo/blob/ddcc2a6c4eabc74b6b5021da05ff67d6d38ccf68/examples/tts/fastpitch.py)). 
- `extract_mels.py` can be used to extract `mels_teacher`, `attentions`, `durations`, `pitch_mel`, `pitch_char` and `pitch_trichar`.

Still there is no information on how to compute `pitch_avg`, `pitch_std`, `pitch_fmin` and `pitch_fmax` (required by [fastpitch_fintune.py](https://github.com/NVIDIA/NeMo/blob/ddcc2a6c4eabc74b6b5021da05ff67d6d38ccf68/examples/tts/fastpitch_finetune.py)). The wording in the pynote mentioned by @blisc gives an impression that the values are empirical rather than precomputed. If not, are these merely the average, standard deviation, minimum and maximum values of the precomputed pitch vectors of a specific speaker (like in https://github.com/NVIDIA/DeepLearningExamples/blob/afe00719cad16281130dc1c260f4575d0f9a8656/PyTorch/SpeechSynthesis/FastPitch/extract_mels.py#L168)? In the later case, which one? Of `pitch` computed via `prepare_dataset.py`? Or `pitch_mel`, `pitch_char`, `pitch_trichar` as computed via `extract_mels.py`? Which of the three pitch versions (mel, char, trichar)?

**Comment by Oktai15** [comment]
Hello, @itzsimpl!

Thanks for your feedback. We are already on the way to fix the problem with different datasets. All TTS models will use only `nemo.collections.tts.torch.data.TTSDataset`, `nemo.collections.tts.torch.data.VocoderDataset `or classes which are inherited from them. 

 Pre-processing also will be standardized. Take a look at already mentioned script: https://github.com/NVIDIA/NeMo/blob/main/scripts/dataset_processing/tts/extract_sup_data.py. We will also update documentation soon where we cover dataset downloading & preprocessing.

**Comment by itzsimpl** [comment]
Glad to hear that, and thank you for an otherwise awesome tool. 

Since you're working on standardization, please make sure to make configurable and evident the use of `parsers` and anything else that might be language or character set (ascii vs utf-8) dependent. As for all authors working with non-English data these can lead to big issues and heavy headaches, especially if they are hardcoded to `en` someplace deep inside the code base. One such prime example being [fastpitch.py](https://github.com/NVIDIA/NeMo/blob/ddcc2a6c4eabc74b6b5021da05ff67d6d38ccf68/nemo/collections/tts/models/fastpitch.py#L184) itself.

**Comment by itzsimpl** [comment]
At first glance this looks much better, of course more comments for some parameters can always improve things even further. For example the `labels` parameter is gone. I assume it is extracted from the model of which one starts off and is kept the same. Are there any parameters that depend on the size of this set (e.g. `symbols_embedding_dim`)? 

Is this config file however reserved only for FastPitch with Aligner (where you start off of a pretrained model)? Note that the `parser` that I earlier pointed to, is created when running FastPitch without Aligner. How would one then start off, when working on a non-English language. A language for which there is no pre-existing FastPitch model. What would be the best way to start?

**Comment by Oktai15** [comment]
@itzsimpl You don't need pretrained model to train FastPitch + Aligner, because Aligner will be trained in the same time as FastPitch.

**Comment by itzsimpl** [comment]
So how does one specify `labels` in case of https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/fastpitch_align_v1.05.yaml? Is it autoextracted? In this specific case (`en`) based on ARPAbet (from the CMUdict)? 

As you probably know by now, I am working with a non-English dataset, hence I am trying to understand the code base, see where everything is language dependant, and try to guess how to make it work with my own language. So pardon me, if I raise too many questions.

**Comment by Oktai15** [comment]
@itzsimpl, take a look at all classes in config. For example, https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/tts_tokenizers.py#L228. Take a look at FastPitch too, for example, this line gets the vocabulary size: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/fastpitch.py#L74

**Comment by mabajec** [comment]
> 

@Oktai15, in the config file you provide as part of the tutorial, one of the parameters is "model.symbols_embedding_dim". In the example its value is 384. Is this value anyhow dependent on the vocabulary size? If yes, how does one calculate it assuming that we train with AudioToCharWithPriorAndPitchDataset class but for a different grapheme based vocab?

**Comment by Oktai15** [comment]
@mabajec it is not necessarily to change `symbols_embedding_dim` even you changed the vocabulary size.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+    "        if stage == \"train\":\n",
+    "            pitch_tensor = torch.cat(pitch_list)\n",
+    "            pitch_mean, pitch_std = pitch_tensor.mean().item(), pitch_tensor.std().item()\n",
+    "            pitch_min, pitch_max = pitch_tensor.min().item(), pitch_tensor.max().item()\n",
+    "            \n",
+    "    return pitch_mean, pitch_std, pitch_min, pitch_max"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (redoctopus) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:404]: Could you add a quick description for this code block? e.g. to tell users to modify "sup_data_folder" to the directory in which they want the supplementary data to be saved, like in the later FastPitch section.

**Comment by mabajec** [comment]
@Oktai15 and everybody, I have a question regarding FastPitch. The config that I'm using is based on fastpitch_align_v1.05.yaml - TTSDataset. But I am wondering which parameters should I observe in tensorboard, since loss is constantly NaN and val loss a constant. The train pitch also seems surprisingly high. To me, it looks there is no learning at all. Does anyone have similar experience?

image
image

**Comment by subhankar-ghosh** [comment]
Hey @mabajec ,

I was also getting very high pitch loss and it was due to a typo in the config. Can you check your config. There was a typo in the config `fastpitch_align_v1.05.yaml`. train_ds.dataset.pitch_avg -> train_ds.dataset.pitch_mean 

This is the merged PR related to this issue: https://github.com/NVIDIA/NeMo/pull/3593

**Comment by mabajec** [comment]
Thanks @subhankar-ghosh for the hint, this might be it.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "82d93859",
+   "metadata": {},
+   "source": [
+    "### FastPitch\n",
+    "\n",
+    "FastPitch is  a fully-parallel text-to-speech model based on FastSpeech, conditioned on fundamental frequency contours. The model predicts pitch contours during inference. Conditioning on frequency contours improves the overall quality of synthesized speech. For more details about model, please refer to the original [paper](https://arxiv.org/abs/2006.06873).\n",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:164]: This text is old, because NeMo's FastPitch implementation has Aligner and you didn't mention it, please, add this information here. And also add link to original implementation (https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechSynthesis/FastPitch), because Aligner integration was created by not NeMo team and information about Aligner is not covered in FastPitch paper.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+    "\n",
+    "FastPitch like most NeMo models is defined as a LightningModule.\n",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:166]: Remove it.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+    "\n",
+    "Let's take a look using NeMo's pretrained model and how to use it to generate spectrograms."
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:168]: We don't need it, because it is just repeated text from Mixer-TTS.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb (hunk 4 above)
Comment (Oktai15) [reply tutorials/tts/MixerTTS_FastPitch_Training.ipynb:164]: Move information about FastPitch after this text about Mixer-TTS:
```
Mixer-TTS is non-autoregressive model for mel-spectrogram generation. The model is based on the MLP-Mixer architecture adapted for speech synthesis. It contains pitch and duration predictors, with the latter being trained with an unsupervised TTS alignment framework. For more details about model, please refer to the original [paper](https://arxiv.org/abs/2110.03584).

Mixer-TTS like most NeMo models is defined as a LightningModule, allowing for easy training via PyTorch Lightning, and parameterized by a configuration, currently defined via a yaml file and loading using Hydra.
```

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+   "source": [
+    "We will be using `nemo.collections.tts.torch.data.TTSDataset` for the entire data pipeline. \n",
+    "\n",
+    "In the Mixer-TTS pipeline we used a phoneme based tokenizer but in the FastPitch training pipeline we would like to demonstrate a character based tokenizer `nemo.collections.tts.torch.tts_tokenizers.EnglishCharsTokenizer`. Unlike Phoneme based tokenizer, `EnglishCharsTokenizer` does not need a phoneme dictionary or a heteronym dictionary. We will be using the same `nemo_text_processing.text_normalization.normalize.Normalizer` for normalizing the text as used in the Mixer-TTS example.\n",
+    "\n",
+    "Since this process does not need any additional data we just download the raw waveform audio files and the manifests."
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:470]: Why do you download it again? It was already done in Mixer-TTS part. Remove it.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+   "execution_count": null,
+   "id": "448a8e94",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!mkdir -p tts_dataset_files && cd tts_dataset_files \\\n",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:487]: Why do you download it again? It was already done in Mixer-TTS part. Remove it.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "4ef5bdb7",
+   "metadata": {},
+   "source": [
+    "Let's define the above mentioned Normalizer and Tokenizer"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:501]: You don't need to import it again, it was already in Mixer-TTS part. Remove it.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 10]
```diff
@@ -0,0 +1,704 @@
...
+    "# text tokenizer\n",
+    "text_tokenizer = EnglishCharsTokenizer()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "55aa4475",
+   "metadata": {},
+   "source": [
+    "Just like in Mixer-TTS we will need to extract pitch for every audio, estimate pitch statistics (mean and std) and pre-calculate alignment prior matrices. The extracted pitch and pre-calculated alignment prior matrices will be stored in `sup_data_path`"
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:540]: Add `.` in the end of text.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 11]
```diff
@@ -0,0 +1,704 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dc446e97",
+   "metadata": {},
+   "source": [
+    "Let's look at some of the options in the training command:\n",
+    " \n",
+    "- *sup_data_types* - List of types of supplementary data to be provided to the model\n",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:656]: You already added description about `sup_data_types`, remove it.

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb [hunk 12]
```diff
@@ -0,0 +1,704 @@
...
+    "\n",
+    "- *sub_data_path* - Path to the directory that contains the supplementary data\n",
```
Comment (Oktai15) [inline tutorials/tts/MixerTTS_FastPitch_Training.ipynb:658]: You already added description about `sub_data_path `, remove it.

**Review by Oktai15 — CHANGES_REQUESTED** [review]

File: tutorials/tts/MixerTTS_FastPitch_Training.ipynb (hunk 9 above)
Comment (subhankar-ghosh) [reply tutorials/tts/MixerTTS_FastPitch_Training.ipynb:468]: Done.

**Review by Oktai15 — APPROVED** [review]
